src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs

```diff
@@ -118,28 +122,29 @@ public async ValueTask SynchronizeProjectAssetsAsync(ProjectStateChecksums proje
 
         using (Logger.LogBlock(FunctionId.AssetService_SynchronizeAssetsAsync, Checksum.GetChecksumsLogInfo, checksums, cancellationToken))
         {
-            using var _1 = ArrayBuilder<Checksum>.GetInstance(checksums.Count, out var missingChecksums);
+            using var missingChecksumsObject = s_checksumListPool.GetPooledObject();
+            var missingChecksums = missingChecksumsObject.Object;
+
             foreach (var checksum in checksums)
             {
                 if (!_assetCache.TryGetAsset<object>(checksum, out _))
                     missingChecksums.Add(checksum);
             }
 
-            var checksumsArray = missingChecksums.ToImmutableAndClear();
-            var assets = await RequestAssetsAsync(assetHint, checksumsArray, cancellationToken).ConfigureAwait(false);
+            var assets = await RequestAssetsAsync(assetHint, missingChecksums, cancellationToken).ConfigureAwait(false);
 
-            Contract.ThrowIfTrue(checksumsArray.Length != assets.Length);
+            Contract.ThrowIfTrue(missingChecksums.Count != assets.Length);
 
             for (int i = 0, n = assets.Length; i < n; i++)
-                _assetCache.GetOrAdd(checksumsArray[i], assets[i]);
+                _assetCache.GetOrAdd(missingChecksums[i], assets[i]);
         }
     }
 
     private async Task<ImmutableArray<object>> RequestAssetsAsync(
-        AssetHint assetHint, ImmutableArray<Checksum> checksums, CancellationToken cancellationToken)
+        AssetHint assetHint, SegmentedList<Checksum> checksums, CancellationToken cancellationToken)
     {
-        Contract.ThrowIfTrue(checksums.Contains(Checksum.Null));
-        if (checksums.Length == 0)
+        Contract.ThrowIfTrue((checksums as ICollection<Checksum>).Contains(Checksum.Null));
```

Comment:
`Contract.ThrowIfTrue` is not a debug-only operation. Are we sure (for `ImmutableSegmentedList<T>`) that this isn't introducing a boxing operation?

Reply:
I don't remember why I needed to do this originally, but it looks like SegmentedList implements Contains directly, so this cast shouldn't be necessary any more.